src/threading/thread.cpp

```diff
@@ -0,0 +1,350 @@
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+*/
+
+#include "threading/thread.h"
+#include "threading/mutex_auto_lock.h"
+
+#if __cplusplus >= 201103L
+	#define THREAD_HANDLE thread->native_handle()
+	#include <chrono>
+#else
+	#define THREAD_HANDLE thread
+	#define UNUSED(expr) do { (void)(expr); } while (0)
+# ifdef _WIN32
+#  ifndef _WIN32_WCE
+	#include <process.h>
+#  endif
+# else
+	#include <ctime>
+	#include <cassert>
+	#include <cstdlib>
+	#include <sys/time.h>
+
+	// For getNumberOfProcessors
+	#include <unistd.h>
+#  if defined(__FreeBSD__) || defined(__APPLE__)
+	#include <sys/types.h>
+	#include <sys/sysctl.h>
+#  elif defined(_GNU_SOURCE)
+	#include <sys/sysinfo.h>
+#  endif
+# endif
+#endif
+
+
+// For setName
+#if defined(linux) || defined(__linux)
+	#include <sys/prctl.h>
+#elif defined(__FreeBSD__) || defined(__OpenBSD__)
+	#include <pthread_np.h>
+#elif defined(_MSC_VER)
+	struct THREADNAME_INFO {
+		DWORD dwType; // Must be 0x1000
+		LPCSTR szName; // Pointer to name (in user addr space)
+		DWORD dwThreadID; // Thread ID (-1=caller thread)
+		DWORD dwFlags; // Reserved for future use, must be zero
+	};
+#endif
+
+// For bindToProcessor
+#if __FreeBSD_version >= 702106
+	typedef cpuset_t cpu_set_t;
+#elif defined(__linux) || defined(linux)
+	#include <sched.h>
+#elif defined(__sun) || defined(sun)
+	#include <sys/types.h>
+	#include <sys/processor.h>
+	#include <sys/procset.h>
+#elif defined(_AIX)
+	#include <sys/processor.h>
+#elif defined(__APPLE__)
+	#include <mach/mach_init.h>
+	#include <mach/thread_policy.h>
+#endif
+
+
+Thread::Thread(const std::string &name) :
+	name(name),
+	retval(NULL),
+	request_stop(false),
+	running(false)
+#if __cplusplus >= 201103L
+	, thread(NULL)
+#elif !defined(_WIN32)
+	, started(false)
+#endif
+{}
+
+
+void Thread::wait()
+{
+#if __cplusplus >= 201103L
+	if (!thread || !thread->joinable()) return;
+	thread->join();
+#else
+# ifdef _WIN32
+	if (!running) return;
+	WaitForSingleObject(thread, INFINITE);
+# else // pthread
+	void *status;
+	if (!started) return;
+	int ret = pthread_join(thread, &status);
+	assert(!ret);
+	UNUSED(ret);
+	started = false;
+# endif
+#endif
+}
+
+
+bool Thread::start()
+{
+	if (running) return false;
+	request_stop = false;
+
+#if __cplusplus >= 201103L
+	MutexAutoLock l(continue_mutex);
+	thread = new std::thread(theThread, this);
+#else
+# ifdef _WIN32
+	MutexAutoLock l(continue_mutex);
+#  ifdef _WIN32_WCE
+	thread = CreateThread(NULL, 0, theThread, this, 0, &thread_id);
```

Comment:
Ah, that's true.  That THREAD_HANDLE is sort of confusing.  What I would personally do here is make a getter function that varies is defined separately with each preprocessor case.